db/view/view.cc

```diff
@@ -1576,6 +1576,16 @@ bool needs_static_row(const mutation_partition& mp, const std::vector<view_and_b
     return mp.partition_tombstone() || !mp.static_row().empty();
 }
 
+bool am_i_a_replica(const locator::effective_replication_map_ptr& ermp, dht::token token) {
+    // Based on the computation in get_view_natural_endpoint, this is used
+    // to detect the case that we're an "official" replica which is
+    // paired with a view replica and sends view updates to.
+    // For a pending replica, for example, this will return false.
+    const auto my_host_id = ermp->get_token_metadata_ptr()->get_topology().my_host_id();
+    const auto replicas = ermp->get_replicas(token);
+    return std::find(replicas.begin(), replicas.end(), my_host_id) != replicas.end();
+}
+
```

Comment:
hmm we may have to think how to handle it with intra-node migration, since 2 shards on the same node will be written to, during migration, but only 1 will be ready for reads (the source or destination)

I think we can fix it by also checking if ermp->shard_for_reads(token) points to this_shard_id().

Also for symmetry with the code base, I recommend you to rename this function to belongs_to_this_shard(). It's no longer about this node owning this token, but rather node:shard pair.


You can test it by doing intra-node migration too. You can parametrize the migration type, so the test stresses both.

/cc @tgrabiec

Reply:
It seems the current approach of the test doesn't work for intra-node migration, I think it doesn't reach stream_mutation_fragments
Do you know how I can make the migration process wait?

Reply:
I think you can add a injection similar to the one stream_mutation_fragments in storage_service

```
diff --git a/service/storage_service.cc b/service/storage_service.cc
index 8006a6706d..47abcb9737 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -5804,6 +5804,15 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
                                                      tablet, leaving_replica));
             }
             tm = nullptr;
+
+            co_await utils::get_local_injector().inject("intranode_migration_streaming_wait", [this] (auto& handler) -> future<> {
+                rtlogger.info("intranode_migration_streaming: waiting");
+                while (!handler.poll_for_message() && !_async_gate.is_closed()) {
+                    co_await sleep(std::chrono::milliseconds(5));
+                }
+                rtlogger.info("intranode_migration_streaming: released");
+            });
+
             rtlogger.info("Starting intra-node streaming of tablet {} from shard {} to {}", tablet, leaving_replica->shard, pending_replica->shard);
             co_await clone_locally_tablet_storage(tablet, *leaving_replica, *pending_replica);
             rtlogger.info("Finished intra-node streaming of tablet {} from shard {} to {}", tablet, leaving_replica->shard, pending_replica->shard);
```

Reply:
Thanks.
I added the condition checking the shard and extended the test to intra-node migration